vochain/transaction/nonce.go

```diff
@@ -0,0 +1,112 @@
+package transaction
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"go.vocdoni.io/dvote/crypto/ethereum"
+	"go.vocdoni.io/dvote/vochain/transaction/vochaintx"
+	"go.vocdoni.io/proto/build/go/models"
+)
+
+// ExtractNonceAndSender extracts the nonce and sender address from a given Vochain transaction.
+// The function uses the signature of the transaction to derive the sender's public key and subsequently
+// the Ethereum address. The nonce is extracted based on the specific payload type of the transaction.
+// If the transaction does not contain signature or nonce, it returns the default values (nil and 0).
+func (t *TransactionHandler) ExtractNonceAndSender(vtx *vochaintx.Tx) (*common.Address, uint32, error) {
+	type txWithNonce interface {
+		GetNonce() uint32
+	}
+	var txNonce txWithNonce
+	var err error
+	switch payload := vtx.Tx.Payload.(type) {
+	case *models.Tx_NewProcess:
+		ptx := payload.NewProcess
+		if ptx == nil {
+			err = fmt.Errorf("new process payload is nil")
+			break
+		}
+		txNonce = txWithNonce(ptx)
+	case *models.Tx_SetProcess:
+		ptx := payload.SetProcess
+		if ptx == nil {
+			err = fmt.Errorf("set process payload is nil")
+			break
+		}
+		txNonce = txWithNonce(ptx)
+	case *models.Tx_SendTokens:
+		ptx := payload.SendTokens
+		if ptx == nil {
+			err = fmt.Errorf("send tokens payload is nil")
+			break
+		}
+		txNonce = txWithNonce(ptx)
+	case *models.Tx_SetAccount:
+		ptx := payload.SetAccount
+		if ptx == nil {
+			err = fmt.Errorf("set account payload is nil")
+			break
+		}
+		if ptx.Txtype == models.TxType_CREATE_ACCOUNT {
+			// create account tx is a special case where the nonce is not relevant
+			return nil, 0, nil
+		}
+		txNonce = txWithNonce(ptx)
+	case *models.Tx_CollectFaucet:
+		ptx := payload.CollectFaucet
+		if ptx == nil {
+			err = fmt.Errorf("collect faucet payload is nil")
+			break
+		}
+		txNonce = txWithNonce(ptx)
+	case *models.Tx_Vote, *models.Tx_Admin, *models.Tx_MintTokens, *models.Tx_SetKeykeeper,
+		*models.Tx_SetTransactionCosts, *models.Tx_DelSIK, *models.Tx_RegisterKey, *models.Tx_SetSIK,
+		*models.Tx_RegisterSIK:
+		// these tx does not have incremental nonce
+		return nil, 0, nil
+	default:
+		// force panic to detect new tx types
+		panic(fmt.Sprintf("unknown payload type on extract nonce: %T", payload))
```

Comment:
is this a debugging leftover?
then maybe just `log.Warnf`

Reply:
It's in purpose (as the comment says) to detect new tx types which have not been added here. So it actas as a protection for future new txs.

Reply:
but that means if someone broadcasts a transaction of a new type, the whole network is immediately brought down with a panic?

Reply:
That's a good point. But I believe Protobuf won't allow the unmarshal or would set it to default value which is models.Tx_UNKNOWN. It is something to check, for sure.

If the protobuf models contain the txType because it is a new introduced tx type, then the tests should fail.

Reply:
if protobuf "protects" us from a DDOS, that's very nice, but in any case, we should really avoid panics in the codepath that processes input from outside, i bet it would raise questions from any security auditor

Reply:
I have moved the Panic to log.Error(), thanks for the input @altergui